mungegithub/mungers/approval-handler.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mungers
+
+import (
+	"path"
+	"strings"
+
+	"k8s.io/contrib/mungegithub/features"
+	"k8s.io/contrib/mungegithub/github"
+	"k8s.io/contrib/mungegithub/mungers/mungerutil"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+)
+
+const maxDepth = 3
+
+// ApprovalHandler will try to add "approved" label once
+// all files of change has been approved by approvers.
+type ApprovalHandler struct {
+	features *features.Features
+}
+
+func init() {
+	h := &ApprovalHandler{}
+	RegisterMungerOrDie(h)
+}
+
+// Name is the name usable in --pr-mungers
+func (*ApprovalHandler) Name() string { return "approval-handler" }
+
+// RequiredFeatures is a slice of 'features' that must be provided
+func (*ApprovalHandler) RequiredFeatures() []string {
+	return []string{features.RepoFeatureName, features.AliasesFeature}
+}
+
+// Initialize will initialize the munger
+func (h *ApprovalHandler) Initialize(config *github.Config, features *features.Features) error {
+	h.features = features
+	return nil
+}
+
+// EachLoop is called at the start of every munge loop
+func (*ApprovalHandler) EachLoop() error { return nil }
+
+// AddFlags will add any request flags to the cobra `cmd`
+func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
+
+// Munge is the workhorse the will actually make updates to the PR
+// The algorithm goes as:
+// - Initially, we set up approverMap
+//   - Go through all comments after latest commit. If any approver said "/approve", add him to approverMap.
+// - For each file, we see if any approver of this file is in approverMap.
+//   - An approver of a file is defined as:
+//     - It's known that each dir has a list of approvers. (This might not hold true. For usability, current situation is enough.)
+//     - Approver of a dir is also the approver of child dirs.
+//   - We look at top N (default 3) level dir approvers. For example, for file "/a/b/c/d/e", we might search for approver from
+//     "/", "/a/", "/a/b/"
+// - Iff all files has been approved, the bot will add "approved" label.
+func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	if !obj.IsPR() {
+		return
+	}
+	files, err := obj.ListFiles()
+	if err != nil {
+		glog.Errorf("failed to list files in this PR: %v", err)
+		return
+	}
+
+	comments, err := getCommentsAfterLastModified(obj)
```

Comment:
@apelisse in the event that file A is modified, only file A should require re-approval right?  I think this implementation will require all approvers write ```/approve``` after every new commit, if if the file or files they previously approved have not changed at all.  is that right?